src/plexil-adapter/OwInterface.cpp

```diff
@@ -310,6 +308,97 @@ static void handle_joint_fault (Joint joint, int joint_index,
   handle_overtorque (joint, msg->effort[joint_index]);
 }
 
+void OwInterface::systemFaultMessageCallback
+(const  ow_faults::SystemFaults::ConstPtr& msg)
+{
+  // Publish all joint information for visibility to PLEXIL and handle any
+  // system-level fault messages.
+  uint64_t msg_val = msg->value;
+
+  for (auto const& entry : systemErrors){
+    string key = entry.first;
+    uint64_t value = entry.second.first;
+    bool b = entry.second.second;
+
+    if (!b && ((msg_val & value) == value)){
+      ROS_ERROR("SYSTEM ERROR: %s", key.c_str() );
+      systemErrors[key].second = true;
+    }
+    else if (b && ((msg_val & value) != value)){
+      ROS_INFO("RESOLVED SYSTEM ERROR: %s", key.c_str() );
+      systemErrors[key].second = false;
+    }
+  }
+}
+
+void OwInterface::armFaultCallback
+(const  ow_faults::ArmFaults::ConstPtr& msg)
+{
+  // Publish all ARM COMPONENT FAULT information for visibility to PLEXIL and handle any
+  // system-level fault messages.
+  uint32_t msg_val = msg->value;
+  
+  for (auto const& entry : armErrors){
+    string key = entry.first;
+    uint32_t value = entry.second.first;
+    bool b = entry.second.second;
+
+    if (checkFaultMessages("ARM", msg_val, key, value, b)) {
+      armErrors[key].second = !armErrors[key].second;
+    }
+  }
+}
+
+void OwInterface::powerFaultCallback
+(const  ow_faults::PowerFaults::ConstPtr& msg)
+{
+  // Publish all POWER FAULT information for visibility to PLEXIL and handle any
+  // system-level fault messages.
+  uint32_t msg_val = msg->value;
+  
+  for (auto const& entry : powerErrors){
+    string key = entry.first;
+    uint32_t value = entry.second.first;
+    bool b = entry.second.second;
+
+    if (checkFaultMessages("POWER", msg_val, key, value, b)) {
+      powerErrors[key].second = !powerErrors[key].second;
+    }
+  }
+}
+
+void OwInterface::antennaFaultCallback
+(const  ow_faults::PTFaults::ConstPtr& msg)
+{
+  // Publish all PANT TILT ANTENNA information for visibility to PLEXIL and handle any
+  // system-level fault messages.
+  uint32_t msg_val = msg->value;
+  
+  for (auto const& entry : ptErrors){
+    string key = entry.first;
+    uint32_t value = entry.second.first;
+    bool b = entry.second.second;
+
+    if (checkFaultMessages("ANTENNA", msg_val, key, value, b)) {
+      ptErrors[key].second = !ptErrors[key].second;
+    }
+  }
+}
+
+bool OwInterface::checkFaultMessages
+(string fault_component, uint32_t msg_val, string key, uint32_t value, bool b )
```

Comment:
Use templates to generalize the method for different types of error vectors
```suggestion
template <typename T>
bool OwInterface::checkFaultMessages(
  string fault_component, T msg_val, string key, T value, bool flag )
```
You need to change the method declaration too.